pandas/_libs/tslibs/period_conversion.pyx

```diff
@@ -0,0 +1,112 @@
+
+
+daytime_conversion_factors = [[FR_DAY, 1],   [FR_HR, 24],
+                              [FR_MIN, 60],  [FR_SEC, 60],
+                              [FR_MS, 1000], [FR_US, 1000],
+                              [FR_NS, 1000], [0, 0]]
+
+
+cdef int64_t[:, :] daytime_conversion_factor_matrix = np.array(
+    [[<int64_t>val for val in row] for row in
+    # Building on Windows seems to require super-explicit casting
+    [
+    [0, 0, 0, 0, 0, 0, 0,  0,    0,     0,        0,           0,              0],
+    [0, 0, 0, 0, 0, 0, 0,  0,    0,     0,        0,           0,              0],
+    [0, 0, 0, 0, 0, 0, 0,  0,    0,     0,        0,           0,              0],
+    [0, 0, 0, 0, 0, 0, 0,  0,    0,     0,        0,           0,              0],
+    [0, 0, 0, 0, 0, 0, 0,  0,    0,     0,        0,           0,              0],
+    [0, 0, 0, 0, 0, 0, 0,  0,    0,     0,        0,           0,              0],
+    [0, 0, 0, 0, 0, 0, 1, 24, 1440, 86400, 86400000, 86400000000, 86400000000000],
+    [0, 0, 0, 0, 0, 0, 0,  1,   60,  3600,  3600000,  3600000000,  3600000000000],
+    [0, 0, 0, 0, 0, 0, 0,  0,    1,    60,    60000,    60000000,    60000000000],
+    [0, 0, 0, 0, 0, 0, 0,  0,    0,     1,     1000,     1000000,     1000000000],
+    [0, 0, 0, 0, 0, 0, 0,  0,    0,     0,        1,        1000,        1000000],
+    [0, 0, 0, 0, 0, 0, 0,  0,    0,     0,        0,           1,           1000],
+    [0, 0, 0, 0, 0, 0, 0,  0,    0,     0,        0,           0,              1]]],
+    dtype=np.int64)  # noqa
+
+
+
+
+@cython.boundscheck(False)
+cdef int64_t get_daytime_conversion_factor(int from_index, int to_index) nogil:
+    cdef:
+        Py_ssize_t row, col
+
+    row = min_value(from_index, to_index)
+    col = max_value(from_index, to_index)
+    return daytime_conversion_factor_matrix[row][col]
+
+
+cdef inline int max_value(int a, int b) nogil:
+    if a > b:
+        return a
+    return b
+
+
+cdef inline int min_value(int a, int b) nogil:
+    if a < b:
+        return a
+    return b
+
+
+@cython.cdivision
+cdef double get_abs_time(int freq, int64_t date_ordinal,
+                         int64_t ordinal) nogil:
+    cdef:
+        int freq_index, day_index, base_index
+        int64_t per_day, start_ord
+        double unit, result
+
+    if freq <= FR_DAY:
+        return 0
+
+    freq_index = get_freq_group_index(freq)
+    day_index = get_freq_group_index(FR_DAY)
+    base_index = get_freq_group_index(FR_SEC)
+
+    per_day = get_daytime_conversion_factor(day_index, freq_index)
+    unit = get_daytime_conversion_factor(freq_index, base_index)
+
+    if base_index < freq_index:
+        unit = 1 / unit
+
+    start_ord = date_ordinal * per_day
+    result = <double>(unit * (ordinal - start_ord))
+    return result
+
+
+# ----------------------------------------------------------------------
+# Conventions
+
+@cython.cdivision
+cdef inline int get_freq_group_index(int freq) nogil:
```

Comment:
maybe lets make a ``pandas._libs/period/....`` subdir here rather than trying to jam into tslibs (and move period as well)

Reply:
a precursor PR to move period would make this easier

Reply:
This is definitely a mess; unfortunately my attempts to port it piece-meal segfaulted all over the place.

But I think this _does_ belong in tslibs.  A ton of the code moved from period_helper is redundant with np_datetime, ccalendar (if we get rid of the unused JULIAN_CALENDAR branches), and frequencies.

Reply:
@shoyer IIRC you mentioned xarray users needing non-gregorian calendar implementations of... something.  Am I remembering that right?  If so, is there anything useful in the JULIAN_CALENDAR code for period_helper that should be retained instead of ripped out?

Reply:
ok, still make it a subdir, maybe ``pandas/_lib/tslibs/period`` then

Reply:
I don't think the Julian calendar part is worth keeping around for xarray users.

Climate scientists use non-Gregorian calendars (e.g., a "365 day calendar") to simplify keeping track of seasons in climate models. But I don't think the Julian calendar is particularly interesting here.